### ISSUE TITLE
Implement CWG2369

### ISSUE THREAD

**Issue by gonzalobg** [llvm/llvm-project#54440]
From http://wg21.link/cwg2369 , https://gcc.godbolt.org/z/nEqhM5dsG
The changes are already in the draft, starting here (https://eel.is/c++draft/temp.deduct.general#5.sentence-8) till the end of that paragraph.

```
template <class T> struct Z {
  typedef typename T::x xx;
};
template <class T> concept C = requires { typename T::A; };
template <C T> typename Z<T>::xx f(void *, T); // #1
template <class T> void f(int, T);             // #2
struct A {} a;
struct ZZ {
    template <class T, class = typename Z<T>::xx> operator T *();
    operator int();
};
int main() {
    ZZ zz;
    f(1, a);   // OK, deduction fails for #1 because there is no conversion from int to void*
    f(zz, 42); // OK, deduction fails for #1 because C<int> is not satisfied
}
```

compiles with GCC 11 and newer, but fails with clang-trunk and GCC 10 and older because they don't implement the CWG2369.

**Comment by jyknight** [comment]
Xref to an issue users ran into with the GCC implementation, in case it's useful to take into account while implementing in Clang: https://gcc.gnu.org/PR99599 -- though I'm not sure what the proper outcome of that is.

**Comment by cor3ntin** [comment]
I heard of more users who need the implementation of this

**Comment by cor3ntin** [comment]
@zyn0217 Hum, the lambda case might indeed be tricky... it would be great to support at least the non-lambda case, and go from there. Thanks!

**Comment by zyn0217** [comment]
Updates:

I have roughly implemented it through some copy-pasting, and the draft can be found here: https://github.com/zyn0217/llvm-project/commits/cwg-2369/

The sketchy implementation doesn't cope with lambdas because we have to incorporate the instantiated captures before the constraint evaluation, which requires more effort. (I was not entirely sure if there's much value in handling lambdas; AFAICT, this CWG issue is seemingly targeted a better overload resolution?)

Despite that, my approach instantiates all function parameters before going into constraint checking so that we can evaluate them properly. However, this could be improved, as we just need to instantiate parameters "used" by the type constraints.

The codes are not well-organized and have caused some regressions in the tests. (All are related to diagnostics, and at least no crashes were left after some tweaking.)

**Comment by jyknight** [comment]
This change regressed the test-case I mentioned before, in the GCC bug-report. Note that GCC has since addressed that case. (See https://gcc.gnu.org/PR99599, esp comment 1 and comment 18).

That is, Clang now refuses to compile code which it did prior to this commit, and which GCC does (again) support.

Should we implement the same fix to the semantics that GCC implemented?

**Comment by zyn0217** [comment]
> This change regressed the test-case I mentioned before, in the GCC bug-report. Note that GCC has since addressed that case. (See https://gcc.gnu.org/PR99599, esp comment 1 and comment 18).
> 
> That is, Clang now refuses to compile code which it did prior to this commit, and which GCC does (again) support.
> 
> Should we implement the same fix to the semantics that GCC implemented?


@cor3ntin is it https://github.com/llvm/llvm-project/issues/104592 that we discussed months ago?

**Comment by cor3ntin** [comment]
Yes - and given GCC is not conforming I'm not sure we want to do anything @zygoloid 
(afaik the libc++ issue was fixed #104713)

**Comment by jyknight** [comment]
Is GCC's method of avoiding the error actually non-conforming, or simply not _required_?

The fact that there are 7 bugs marked as duplicate against that GCC issue worries me, in that users seem to actually run into this error.

**Comment by zyn0217** [comment]
> Is GCC's method of avoiding the error actually non-conforming, or simply not required?

I think it's more of Not required.

However, the heuristics GCC introduced seem to cause discrepancies between non-template and template code - see #104173. I’m also concerned that introducing them might confuse our users.

That said, I’m not opposed to following GCC if we agree this needs to be resolved before CWG says something. We just need more opinions.

**Comment by frederick-vs-ja** [comment]
@zyn0217 Looks like that we should close as #122423 is landed.
